### PR TITLE
Knex should be implementation detail of Store

### PR DESCRIPTION
**Pull request by snario**
I think we agree that Knex is an implementation detail of a PostgreSQL specific implementation of a Store, thus it should not be known by a Wallet, especially if we intend on re-using that code in the browser.

### CODE REVIEW

**Comment by andrewgordstewart** [comment]
I think this should go even further: it should have models that are bound to its knex.

**Comment by snario** [comment]
Something introduced an unhandled promise in the server wallet tests... not sure what.

**Comment by andrewgordstewart** [comment]
> Something introduced an unhandled promise in the server wallet tests... not sure what.

Each time you construct a store, you construct a knex that must be destroyed for jest to exit.

You could require a knex instance by the store's constructor.

**Comment by snario** [comment]
> I think this should go even further: the store should have models that are bound to its knex.

Ah I see! I think the store should probably create its own connection

**Comment by andrewgordstewart** [comment]
> > I think this should go even further: the store should have models that are bound to its knex.
> 
> Ah I see! I think the store should probably create its own connection

I sort of agree, but then every store that is constructed during a test suite needs to be destroyed. I don't particularly like that tradeoff.

**Comment by snario** [comment]
> I sort of agree, but then every store that is constructed during a test suite needs to be destroyed. I don't particularly like that tradeoff.

Since a store is basically a wrapper around a database connection, I don't think it's that unreasonable really. And presently there are only [five occurrences](https://github.com/statechannels/statechannels/pull/2580/commits/d6bdde2c9517cca54a5bbbf0efd7887ad2a4de92), which could be bundled up in a setup and teardown script if we desired to simplify it, or just a single jest test.

**Comment by snario** [comment]
**ci/circleci: server-wallet-profiling** failure related to https://github.com/statechannels/statechannels/pull/2566

File: packages/server-wallet/src/wallet/__test__/integration/create-channel.test.ts [hunk 1]
```diff
@@ -80,6 +80,4 @@ describe('happy path', () => {
 });
 
 it("doesn't create a channel if it doesn't have a signing wallet", () =>
-  expect(w.createChannel(createChannelArgs())).rejects.toThrow(
-    'null value in column "signing_address"'
-  ));
+  expect(w.createChannel(createChannelArgs())).rejects.toThrow('Not in channel'));
```
Comment (geoknee) [inline packages/server-wallet/src/wallet/__test__/integration/create-channel.test.ts:83]: Why this change?

File: packages/server-wallet/src/wallet/index.ts [hunk 2]
```diff
@@ -162,95 +164,66 @@ export class Wallet implements WalletInterface {
...
-              objectives: [
-                {
-                  participants,
-                  type: 'CreateChannel',
-                  data: {
-                    signedState: params.data.signedStates[0],
-                    fundingStrategy: 'Direct',
-                  },
+      return {
+        method: 'MessageQueued' as const,
```
Comment (geoknee) [inline packages/server-wallet/src/wallet/index.ts:207]: I see `as const` sprinkled around a fair bit. Do you know what is it doing for us?

File: packages/server-wallet/src/wallet/store.ts [hunk 3]
```diff
@@ -299,6 +314,40 @@ export class Store {
...
+    channelId: string,
+    fromAmount: Uint256,
+    assetHolderAddress: Address
+  ): Promise<void> {
+    await Funding.updateFunding(this.knex, channelId, fromAmount, assetHolderAddress);
+  }
+
+  async nextNonce(signingAddresses: Address[]): Promise<number> {
+    return await Nonce.next(this.knex, signingAddresses);
+  }
```
Comment (geoknee) [inline packages/server-wallet/src/wallet/store.ts:350]: Why are these additions in scope for this PR?

File: packages/server-wallet/src/wallet/__test__/lock-app.test.ts [hunk 4]
```diff
@@ -8,13 +8,19 @@ import {Store} from '../store';
...
 jest.setTimeout(10_000);
 
-const store = new Store(defaultConfig.timingMetrics, defaultConfig.skipEvmValidation);
+const store = new Store(
+  extractDBConfigFromServerWalletConfig(defaultConfig),
+  defaultConfig.timingMetrics,
+  defaultConfig.skipEvmValidation
+);
+
+afterAll(async () => store.knex.destroy());
```
Comment (geoknee) [inline packages/server-wallet/src/wallet/__test__/lock-app.test.ts:23]: Suggestion: the `store` could have a `.destroy()` method that destroys its `knex`? The pattern being, a `wallet` destroys its direct descendants (`store`) and a `store` destorys its direct descendants (`knex`).

**Review by geoknee — APPROVED** [review]

File: packages/server-wallet/src/wallet/__test__/integration/create-channel.test.ts (hunk 1 above)
Comment (snario) [reply packages/server-wallet/src/wallet/__test__/integration/create-channel.test.ts:83]: The create channel method throws a `StoreError` now instead of raising a Knex error

File: packages/server-wallet/src/wallet/index.ts (hunk 2 above)
Comment (snario) [reply packages/server-wallet/src/wallet/index.ts:207]: No idea 🤷 I think @kerzhner has been adding these

File: packages/server-wallet/src/wallet/store.ts (hunk 3 above)
Comment (snario) [reply packages/server-wallet/src/wallet/store.ts:350]: The wallet was in control of database models, which relied on a database connection, which is now contained within the store, so now the database models are also inside the store.

File: packages/server-wallet/src/wallet/__test__/lock-app.test.ts (hunk 4 above)
Comment (snario) [reply packages/server-wallet/src/wallet/__test__/lock-app.test.ts:23]: SGTM, will update.

Comment (snario) [reply packages/server-wallet/src/wallet/__test__/lock-app.test.ts]: 520a43d68

File: packages/server-wallet/src/wallet/store.ts (hunk 3 above)
Comment (snario) [reply packages/server-wallet/src/wallet/store.ts:459]: e.g., 
image

File: packages/server-wallet/src/wallet/index.ts (hunk 2 above)
Comment (andrewgordstewart) [reply packages/server-wallet/src/wallet/index.ts:208]: It narrows the type from `string` to a specific string type

File: packages/server-wallet/src/wallet/store.ts [hunk 5]
```diff
@@ -94,6 +94,10 @@ export class Store {
     }
   }
 
+  async closeDatabaseConnection(): Promise<void> {
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/wallet/store.ts:97]: This could probably just be named `destroy`:
- there may be multiple things that the store needs to clean up
- in a world where we have multiple implementations of a `Store` interface, we might not have a `Database`

Also, (extremely small) nit:
- database connections are ephemeral. They are created on demand, and the number can fluctuate depending on the `knex` config as well as both the wallet load (when not under stress, connections are closed after a timeout) and the DB load (the DB may forcibly close a connection for a variety of reasons)
- knex (well, `tarn`, under the hood) maintains a dynamic pool of connections
- `knex.destroy()` destroys the _pool_ of connections, and signals to `knex` not to start any new ones.

File: packages/server-wallet/src/wallet/store.ts [hunk 6]
```diff
@@ -62,7 +65,19 @@ const throwMissingChannel: MissingAppHandler<any> = (channelId: string) => {
 
 // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
 export class Store {
-  constructor(readonly timingMetrics: boolean, readonly skipEvmValidation: boolean) {
+  knex: Knex;
+
+  constructor(
+    readonly knexOrConfig: Knex | Knex.Config,
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/wallet/store.ts:71]: I really think this should _insist_ on accepting a `Knex`.

We are now living in a world with a new "gotcha":
1. Say I create a `new Store` with a config object instead of a `Knex` instance.
2. Then, when I run my test, I will notice that jest doesn't exit.
3. I have to then learn that `Store` constructed a connection pool, and that pool may contain open connections, and that prevents the jest process from exiting.

There is a significant cost to (3).
A. Multiple people on this team have had to learn this the hard way, so this issue has already cost at least a full developer day.
B. When they do learn it, they need to maintain some additional overhead: "should I pass in a config? do I need to destroy the pool?"

I fail to see any benefit in the short-term of doing (1). Any benefit should be clearly stated, and (IMO) planned into our work towards consolidating the server and browser wallets.

File: packages/server-wallet/src/wallet/store.ts (hunk 5 above)
Comment (snario) [reply packages/server-wallet/src/wallet/store.ts:97]: https://github.com/statechannels/statechannels/pull/2602

File: packages/server-wallet/src/wallet/store.ts (hunk 6 above)
Comment (snario) [reply packages/server-wallet/src/wallet/store.ts:71]: https://github.com/statechannels/statechannels/pull/2602
